ProjFS.Mac/PrjFSKextTests/VirtualizationRootsTests.mm

```diff
@@ -0,0 +1,265 @@
+#include "../PrjFSKext/VirtualizationRoots.hpp"
+#include "../PrjFSKext/VirtualizationRootsTestable.hpp"
+#include "../PrjFSKext/PrjFSProviderUserClient.hpp"
+#include "../PrjFSKext/kernel-header-wrappers/mount.h"
+#include "../PrjFSKext/Memory.hpp"
+#include "../PrjFSKext/Locks.hpp"
+#include "../PrjFSKext/VnodeUtilities.hpp"
+#include "../PrjFSKext/KextLog.hpp"
+#include "KextMockUtilities.hpp"
+#include "MockVnodeAndMount.hpp"
+
+#import <XCTest/XCTest.h>
+#include <vector>
+#include <string>
+
+class PrjFSProviderUserClient
+{
+};
+
+
+void ProviderUserClient_UpdatePathProperty(PrjFSProviderUserClient* userClient, const char* providerPath)
+{
+    MockCalls::RecordFunctionCall(ProviderUserClient_UpdatePathProperty, userClient, providerPath);
+}
+
+
+@interface VirtualizationRootsTests : XCTestCase
+
+@end
+
+@implementation VirtualizationRootsTests
+{
+    PrjFSProviderUserClient dummyClient;
+    pid_t dummyClientPid;
+    MountPointer testMountPoint;
+}
+
+- (void)setUp
+{
+    srand(0);
+    self->dummyClientPid = 100;
+    
+    // This is roughly what "real" fsids look like
+    fsid_t testMountFsid = { (1 << 24) | (rand() % 32), (rand() % 16) };
+    
+    // HFS+ inodes are < UINT32_MAX (APFS' are 64-bit)
+    uint64_t testMountInitialInode = rand();
+    
+    testMountPoint = mount::Create("hfs", testMountFsid, testMountInitialInode);
+    
+    kern_return_t initResult = VirtualizationRoots_Init();
+    XCTAssertEqual(initResult, KERN_SUCCESS);
+}
+
+- (void)tearDown
+{
+    VirtualizationRoots_Cleanup();
+
+    MockVnodes_CheckAndClear();
+    MockCalls::Clear();
+}
+
+
+- (void)testRegisterProviderForPath_NonexistentPath
+{
+    const char* path = "/Users/test/code/Does/Not/Exist";
+    VirtualizationRootResult result = VirtualizationRoot_RegisterProviderForPath(&self->dummyClient, self->dummyClientPid, path);
+    XCTAssertEqual(result.error, ENOENT);
+    XCTAssertFalse(VirtualizationRoot_IsValidRootHandle(result.root));
+    
+    XCTAssertFalse(MockCalls::DidCallFunction(vfs_setauthcache_ttl));
+}
+
+- (void)testRegisterProviderForPath_NonDirectoryPath
+{
+    const char* path = "/Users/test/code/NotADirectory.cpp";
+    
+    VnodePointer vnode = vnode::Create(self->testMountPoint, path);
+    
+    VirtualizationRootResult result = VirtualizationRoot_RegisterProviderForPath(&self->dummyClient, self->dummyClientPid, path);
+    XCTAssertEqual(result.error, ENOTDIR);
+    XCTAssertFalse(VirtualizationRoot_IsValidRootHandle(result.root));
+
+    XCTAssertFalse(MockCalls::DidCallFunction(vfs_setauthcache_ttl));
+}
+
+- (void)testRegisterProviderForPath_DisallowedFileSystem
+{
+    const char* path = "/Volumes/USBStick/repo";
+    
+    fsid_t fatFsid = self->testMountPoint->GetFsid();
+    fatFsid.val[1]++;
+    MountPointer fatMount(mount::Create("msdos", fatFsid, rand()));
+    
+    VnodePointer vnode = vnode::Create(fatMount, path);
```

Comment:
Should this `vnode` be created as a `VDIR`?

Reply:
Yes definitely, thanks!